addons/godot-next/nodes/behaviors.gd

```diff
@@ -0,0 +1,178 @@
+class_name Behaviors
+
+##### PROPERTIES #####
+
+var _behaviors: ResourceSet
+
+var _callbacks: Dictionary = {
+	"_ready" : {},
+	"_process" : {},
+	"_physics_process" : {},
+	"_input" : {},
+	"_unhandled_input" : {},
+	"_unhandled_key_input" : {}
+}
+
+##### NOTIFICATIONS #####
+
+func _set(p_property: String, p_value) -> bool:
+	if _behaviors.can_contain(p_property):
+		#warning-ignore:return_value_discarded
+		_behaviors.set_element(p_property, p_value)
+		property_list_changed_notify()
+		return true
+	return false
+
+func _get(p_property: String):
+	if _behaviors.can_contain(p_property):
+		return _behaviors.get_element(p_property)
+	return null
+
+func _get_property_list() -> Array:
+	if not _behaviors:
+		_behaviors = ResourceSet.new()
+	_behaviors.setup("_behaviors", Behavior)
+	var list = [];
+	list += [ PropertyInfo.new_dictionary("_behaviors", 0, "", PROPERTY_USAGE_STORAGE).to_dict() ];
+	list += _behaviors.get_collection_property_list()
+	return list
+
+func _ready() -> void:
+	if not Engine.editor_hint:
+		var behaviors = _behaviors.get_data();
+		for behavior in behaviors.values():
+			_initialize_behavior(behavior)
+		
+		for behavior in _callbacks["_ready"]:
+			behavior._ready()
+
+func _process(delta: float) -> void:
+	for behavior in _callbacks["_process"]:
+		if behavior.enabled:
+			behavior._process(delta)
+
+func _physics_process(delta: float) -> void:
+	for behavior in _callbacks["_physics_process"]:
+		if behavior.enabled:
+			behavior._process(delta)
+
+func _input(event: InputEvent) -> void:
+	for behavior in _callbacks["_input"]:
+		if behavior.enabled:
+			behavior._input(event)
+
+func _unhandled_input(event: InputEvent) -> void:
+	for behavior in _callbacks["_unhandled_input"]:
+		if behavior.enabled:
+			behavior._unhandled_input(event)
+
+func _unhandled_key_input(event: InputEventKey) -> void:
+	for behavior in _callbacks["_unhandled_key_input"]:
+		if behavior.enabled:
+			behavior._process(event)
+
+##### VIRTUALS #####
+
+func _add_element(p_variable: String, p_script: Script) -> void:
+	print(p_variable)
+	var collection: ResourceCollection = get(p_variable)
+	collection.add_element(p_script)
+
+##### PUBLIC METHODS #####
+
+func add_behavior(p_type: Script) -> Behavior:
+	var behaviors = _behaviors.get_data();
+	
+	var ct = ClassType.new(p_type)
+	if not ct.type_is(Behavior):
+		return null
+	if has_behavior(p_type):
+		return get_behavior(p_type)
+		
+	var behavior: Behavior = p_type.new()
+	var behavior_name: String = ct.get_script_class()
+	
+	behaviors[behavior_name] = behavior
+	_initialize_behavior(behavior)
+	
+	return behavior
+
+func get_behavior(p_type: Script) -> Behavior:
+	var behaviors = _behaviors.get_data();
+	
+	var ct = ClassType.new(p_type)
+	var type_name: String = ct.get_script_class()
+	if behaviors.has(type_name):
+		return behaviors[type_name]
+	return null
+
+func has_behavior(p_type: Script) -> bool:
+	var behaviors = _behaviors.get_data();
+	
+	var ct = ClassType.new(p_type)
```

Comment:
for speed's sake, (over-optimization) you can also just do `return behaviors.has(ClassType.new(p_type).get_script_class())`.